google-services-plugin/src/main/kotlin/com/google/gms/googleservices/GoogleServicesPlugin.kt

```diff
@@ -0,0 +1,153 @@
+                    handleVariant(variant, project)
+                }
+            }
+        }
+        project.pluginManager.withPlugin("com.android.dynamic-feature") {
+            pluginApplied = true
+            project.extensions.configure(DynamicFeatureAndroidComponentsExtension::class.java) {
+                it.registerSourceType(SOURCE_TYPE)
+                it.onVariants { variant ->
+                    handleVariant(variant, project)
+                }
+            }
+        }
+
+        project.afterEvaluate {
+            if (pluginApplied) {
+                return@afterEvaluate
+            }
+            project.logger.error(
+                "The google-services Gradle plugin needs to be applied on a project with " +
+                        "com.android.application or com.android.dynamic-feature."
+            )
+        }
+    }
+
+    private fun <T> handleVariant(
+        variant: T,
+        project: Project
+    ) where T : Variant, T : GeneratesApk {
+        val jsonToXmlTask = project.tasks.register(
+            "process${variant.name.capitalize()}GoogleServices",
+            GoogleServicesTask::class.java
+        ) {
+            it.missingGoogleServicesStrategy.set(
+                project.extensions.getByType(GoogleServicesPluginConfig::class.java)
+                    .missingGoogleServicesStrategy
+            )
+            it.googleServicesJsonFiles.set(variant.sources.getByName(SOURCE_TYPE).all)
+            it.applicationId.set(variant.applicationId)
+        }
+
+        // TODO: add an AGP version check to this block
+        //  when https://issuetracker.google.com/issues/268192807 is fixed
+        try {
+            variant.sources.getByName(SOURCE_TYPE)
+                .addStaticSourceDirectory(
+                    project
+                        .layout
+                        .projectDirectory
+                        .dir("src/${variant.name}/google-services").toString()
+                )
+        } catch (e: IllegalArgumentException) {
+            // directory doesn't exist, ignore
+        }
+        variant.sources.res?.addGeneratedSourceDirectory(
+            jsonToXmlTask,
+            GoogleServicesTask::outputDirectory
+        )
+    }
+
+    /* Recommended replacement for Kotlin's deprecated capitalize function */
+    private fun String.capitalize(): String =
+        replaceFirstChar {
+            if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString()
+        }
+
+
+    companion object {
+        const val MODULE_GROUP = "com.google.android.gms"
+        const val MODULE_GROUP_FIREBASE = "com.google.firebase"
+        const val MODULE_CORE = "firebase-core"
+        const val MODULE_VERSION = "11.4.2"
+        const val MINIMUM_VERSION = "9.0.0"
+        const val SOURCE_TYPE = "google-services"
+    }
+
+
+    enum class MissingGoogleServicesStrategy {
+        IGNORE, WARN, ERROR
+    }
+
+    open class GoogleServicesPluginConfig {
```

Comment:
You'll wanna use conventions in place of default variables; see [gradle docs](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build).